source/server/connection_handler_impl.h

```diff
@@ -136,40 +138,52 @@ class ConnectionHandlerImpl : public Network::ConnectionHandler,
      */
     void newConnection(Network::ConnectionSocketPtr&& socket);
 
+    ActiveConnections& getOrCreateActiveConnections(uint64_t tag);
+
     ConnectionHandlerImpl& parent_;
     Network::ListenerPtr listener_;
     const std::chrono::milliseconds listener_filters_timeout_;
     const bool continue_on_listener_filters_timeout_;
     std::list<ActiveTcpSocketPtr> sockets_;
-    std::list<ActiveTcpConnectionPtr> connections_;
+    std::unordered_map<uint64_t, ActiveConnectionsPtr> connections_by_tag_;
```

Comment:
I'm pretty sure we can remove tag from the public interfaces, but this is probably the crux of it right here. What is going on here exactly? You have a tag associated with a filter chain so you can remove it, right? Could you either have the map key be the filter chain context as a shared_ptr, or if required to avoid circular dependencies a weak_ptr? I'm a little unclear on the final ownership model here.

Reply:
Yeah, my very early prototype is using raw pointer of as key, I feel tag is more acceptable in terms of the ownership.
Also the good part of guaranteed no ABA issue by any chance.
@mattklein123 I need you suggestion:
Should I explore the ptr based map in this PR, or should I use the tag based, and remove the tag() from filter chain API in the (hopefully short) future?

Reply:
Can we try to just kill tag now? My gut tells me it's not really needed. I'm not sure if the key should be raw pointer, weak pointer, or shared ptr, but one of those sounds more clear to me.

Reply:
replaced by const Network::FilterChain* here